File: mobile/mobile_central/test/services/modules/shipments/app/use_cases_test.dart
import 'package:flutter_test/flutter_test.dart';
import 'package:mobile_central/services/modules/shipments/app/use_cases.dart';
import 'package:mobile_central/services/modules/shipments/domain/entities.dart';
import 'package:mobile_central/services/modules/shipments/domain/ports.dart';
import 'package:mobile_central/shared/types/paginated_response.dart';

// --- Manual Mock ---

class MockShipmentRepository implements IShipmentRepository {
  final List<String> calls = [];

  PaginatedResponse<Shipment>? getShipmentsResult;
  Map<String, dynamic>? quoteShipmentResult;
  Map<String, dynamic>? generateGuideResult;
  Map<String, dynamic>? trackShipmentResult;
  Map<String, dynamic>? cancelShipmentResult;
  Shipment? createShipmentResult;
  List<OriginAddress>? getOriginAddressesResult;
  OriginAddress? createOriginAddressResult;
  OriginAddress? updateOriginAddressResult;

  Exception? errorToThrow;

  GetShipmentsParams? capturedGetShipmentsParams;
  EnvioClickQuoteRequest? capturedQuoteRequest;
  String? capturedTrackingNumber;
  String? capturedCancelId;
  Map<String, dynamic>? capturedCreateData;
  int? capturedBusinessId;
  CreateOriginAddressDTO? capturedCreateOriginData;
  int? capturedUpdateOriginId;
  Map<String, dynamic>? capturedUpdateOriginData;
  int? capturedDeleteOriginId;

  @override
  Future<PaginatedResponse<Shipment>> getShipments(GetShipmentsParams? params) async {
    calls.add('getShipments');
    capturedGetShipmentsParams = params;
    if (errorToThrow != null) throw errorToThrow!;
    return getShipmentsResult!;
  }

  @override
  Future<Map<String, dynamic>> quoteShipment(EnvioClickQuoteRequest req) async {
    calls.add('quoteShipment');
    capturedQuoteRequest = req;
    if (errorToThrow != null) throw errorToThrow!;
    return quoteShipmentResult!;
  }

  @override
  Future<Map<String, dynamic>> generateGuide(EnvioClickQuoteRequest req) async {
    calls.add('generateGuide');
    capturedQuoteRequest = req;
    if (errorToThrow != null) throw errorToThrow!;
    return generateGuideResult!;
  }

  @override
  Future<Map<String, dynamic>> trackShipment(String trackingNumber) async {
    calls.add('trackShipment');
    capturedTrackingNumber = trackingNumber;
    if (errorToThrow != null) throw errorToThrow!;
    return trackShipmentResult!;
  }

  @override
  Future<Map<String, dynamic>> cancelShipment(String id) async {
    calls.add('cancelShipment');
    capturedCancelId = id;
    if (errorToThrow != null) throw errorToThrow!;
    return cancelShipmentResult!;
  }

  @override
  Future<Shipment> createShipment(Map<String, dynamic> data) async {
    calls.add('createShipment');
    capturedCreateData = data;
    if (errorToThrow != null) throw errorToThrow!;
    return createShipmentResult!;
  }

  @override
  Future<List<OriginAddress>> getOriginAddresses({int? businessId}) async {
    calls.add('getOriginAddresses');
    capturedBusinessId = businessId;
    if (errorToThrow != null) throw errorToThrow!;
    return getOriginAddressesResult!;
  }

  @override
  Future<OriginAddress> createOriginAddress(CreateOriginAddressDTO data, {int? businessId}) async {
    calls.add('createOriginAddress');
    capturedCreateOriginData = data;
    capturedBusinessId = businessId;
    if (errorToThrow != null) throw errorToThrow!;
    return createOriginAddressResult!;
  }

  @override
  Future<OriginAddress> updateOriginAddress(int id, Map<String, dynamic> data, {int? businessId}) async {
    calls.add('updateOriginAddress');
    capturedUpdateOriginId = id;
    capturedUpdateOriginData = data;
    capturedBusinessId = businessId;
    if (errorToThrow != null) throw errorToThrow!;
    return updateOriginAddressResult!;
  }

  @override
  Future<void> deleteOriginAddress(int id, {int? businessId}) async {
    calls.add('deleteOriginAddress');
    capturedDeleteOriginId = id;
    capturedBusinessId = businessId;
    if (errorToThrow != null) throw errorToThrow!;
  }
}

// --- Helpers ---

Shipment _makeShipment({int id = 1, String status = 'pending'}) {
  return Shipment(
    id: id, createdAt: '2026-01-01', updatedAt: '2026-01-01',
    status: status, isLastMile: false, isTest: false,
  );
}

OriginAddress _makeOriginAddress({int id = 1, String alias = 'Office'}) {
  return OriginAddress(
    id: id, businessId: 1, alias: alias, company: 'Co', firstName: 'F',
    lastName: 'L', email: 'e', phone: 'p', street: 's', cityDaneCode: 'd',
    city: 'c', state: 's', isDefault: false, createdAt: 'c', updatedAt: 'u',
  );
}

Pagination _makePagination() {
  return Pagination(
    currentPage: 1, perPage: 10, total: 1, lastPage: 1,
    hasNext: false, hasPrev: false,
  );
}

EnvioClickQuoteRequest _makeQuoteRequest() {
  return EnvioClickQuoteRequest(
    description: 'Package',
    contentValue: 50000,
    includeGuideCost: true,
    codPaymentMethod: 'cash',
    packages: [EnvioClickPackage(weight: 1, height: 10, width: 10, length: 10)],
    origin: EnvioClickAddress(address: 'Origin', daneCode: '11001'),
    destination: EnvioClickAddress(address: 'Dest', daneCode: '76001'),
  );
}

// --- Tests ---

void main() {
  late MockShipmentRepository mockRepo;
  late ShipmentUseCases useCases;

  setUp(() {
    mockRepo = MockShipmentRepository();
    useCases = ShipmentUseCases(mockRepo);
  });

  group('getShipments', () {
    test('delegates to repository and returns result', () async {
      final expected = PaginatedResponse<Shipment>(
        data: [_makeShipment()],
        pagination: _makePagination(),
      );
      mockRepo.getShipmentsResult = expected;
      final params = GetShipmentsParams(page: 1, pageSize: 10);

      final result = await useCases.getShipments(params);

      expect(result.data.length, 1);
      expect(mockRepo.calls, ['getShipments']);
      expect(mockRepo.capturedGetShipmentsParams, params);
    });

    test('passes null params to repository', () async {
      mockRepo.getShipmentsResult = PaginatedResponse<Shipment>(
        data: [], pagination: _makePagination(),
      );

      await useCases.getShipments(null);

      expect(mockRepo.capturedGetShipmentsParams, isNull);
    });

    test('propagates repository errors', () async {
      mockRepo.errorToThrow = Exception('Network error');

      expect(() => useCases.getShipments(null), throwsException);
    });
  });

  group('quoteShipment', () {
    test('delegates to repository with correct request', () async {
      final req = _makeQuoteRequest();
      mockRepo.quoteShipmentResult = {'rates': []};

      final result = await useCases.quoteShipment(req);

      expect(result['rates'], isA<List>());
      expect(mockRepo.capturedQuoteRequest, req);
      expect(mockRepo.calls, ['quoteShipment']);
    });

    test('propagates repository errors', () async {
      mockRepo.errorToThrow = Exception('Quote failed');

      expect(() => useCases.quoteShipment(_makeQuoteRequest()), throwsException);
    });
  });

  group('generateGuide', () {
    test('delegates to repository with correct request', () async {
      final req = _makeQuoteRequest();
      mockRepo.generateGuideResult = {'guide_id': 'G-001'};

      final result = await useCases.generateGuide(req);

      expect(result['guide_id'], 'G-001');
      expect(mockRepo.capturedQuoteRequest, req);
      expect(mockRepo.calls, ['generateGuide']);
    });
  });

  group('trackShipment', () {
    test('delegates to repository with correct tracking number', () async {
      mockRepo.trackShipmentResult = {'status': 'in_transit'};

      final result = await useCases.trackShipment('TRK-001');

      expect(result['status'], 'in_transit');
      expect(mockRepo.capturedTrackingNumber, 'TRK-001');
      expect(mockRepo.calls, ['trackShipment']);
    });
  });

  group('cancelShipment', () {
    test('delegates to repository with correct id', () async {
      mockRepo.cancelShipmentResult = {'message': 'cancelled'};

      final result = await useCases.cancelShipment('1');

      expect(result['message'], 'cancelled');
      expect(mockRepo.capturedCancelId, '1');
      expect(mockRepo.calls, ['cancelShipment']);
    });

    test('propagates repository errors', () async {
      mockRepo.errorToThrow = Exception('Cancel failed');

      expect(() => useCases.cancelShipment('1'), throwsException);
    });
  });

  group('createShipment', () {
    test('delegates to repository with correct data', () async {
      final data = {'order_id': 'ord-1', 'carrier': 'SER'};
      mockRepo.createShipmentResult = _makeShipment(id: 99);

      final result = await useCases.createShipment(data);

      expect(result.id, 99);
      expect(mockRepo.capturedCreateData, data);
      expect(mockRepo.calls, ['createShipment']);
    });
  });

  group('getOriginAddresses', () {
    test('delegates to repository with businessId', () async {
      mockRepo.getOriginAddressesResult = [_makeOriginAddress()];

      final result = await useCases.getOriginAddresses(businessId: 5);

      expect(result.length, 1);
      expect(result[0].alias, 'Office');
      expect(mockRepo.capturedBusinessId, 5);
      expect(mockRepo.calls, ['getOriginAddresses']);
    });
  });

  group('createOriginAddress', () {
    test('delegates to repository with correct data', () async {
      final dto = CreateOriginAddressDTO(
        alias: 'New', company: 'Co', firstName: 'F', lastName: 'L',
        email: 'e', phone: 'p', street: 's', cityDaneCode: 'd',
        city: 'c', state: 's',
      );
      mockRepo.createOriginAddressResult = _makeOriginAddress(id: 10, alias: 'New');

      final result = await useCases.createOriginAddress(dto, businessId: 5);

      expect(result.id, 10);
      expect(result.alias, 'New');
      expect(mockRepo.capturedCreateOriginData, dto);
      expect(mockRepo.capturedBusinessId, 5);
      expect(mockRepo.calls, ['createOriginAddress']);
    });
  });

  group('updateOriginAddress', () {
    test('delegates to repository with correct id and data', () async {
      final data = {'alias': 'Updated'};
      mockRepo.updateOriginAddressResult = _makeOriginAddress(id: 5, alias: 'Updated');

      final result = await useCases.updateOriginAddress(5, data, businessId: 3);

      expect(result.id, 5);
      expect(result.alias, 'Updated');
      expect(mockRepo.capturedUpdateOriginId, 5);
      expect(mockRepo.capturedUpdateOriginData, data);
      expect(mockRepo.capturedBusinessId, 3);
      expect(mockRepo.calls, ['updateOriginAddress']);
    });
  });

  group('deleteOriginAddress', () {
    test('delegates to repository with correct id', () async {
      await useCases.deleteOriginAddress(7, businessId: 3);

      expect(mockRepo.capturedDeleteOriginId, 7);
      expect(mockRepo.capturedBusinessId, 3);
      expect(mockRepo.calls, ['deleteOriginAddress']);
    });

    test('propagates repository errors', () async {
      mockRepo.errorToThrow = Exception('Delete failed');

      expect(() => useCases.deleteOriginAddress(7), throwsException);
    });
  });
}

File: front/central/src/services/modules/shipments/app/use-cases.test.ts
import { describe, it, expect, vi, beforeEach } from 'vitest';
import { ShipmentUseCases } from './use-cases';
import { IShipmentRepository } from '../domain/ports';
import {
    Shipment,
    PaginatedResponse,
    EnvioClickQuoteRequest,
    EnvioClickQuoteResponse,
    EnvioClickGenerateResponse,
    EnvioClickTrackingResponse,
    EnvioClickCancelResponse,
    CreateShipmentRequest,
    OriginAddress,
    CreateOriginAddressRequest,
    UpdateOriginAddressRequest,
} from '../domain/types';

// -----------------------------------------------------------------
// Helpers: datos de prueba reutilizables
// -----------------------------------------------------------------

const makeShipment = (overrides: Partial<Shipment> = {}): Shipment => ({
    id: 1,
    created_at: '2026-03-01T00:00:00Z',
    updated_at: '2026-03-01T00:00:00Z',
    status: 'pending',
    is_last_mile: false,
    is_test: false,
    ...overrides,
});

const makeOriginAddress = (overrides: Partial<OriginAddress> = {}): OriginAddress => ({
    id: 1,
    business_id: 1,
    alias: 'Bodega Principal',
    company: 'Mi Empresa',
    first_name: 'Juan',
    last_name: 'Perez',
    email: 'juan@test.com',
    phone: '3001234567',
    street: 'Calle 10 #5-20',
    city_dane_code: '11001',
    city: 'Bogota',
    state: 'Cundinamarca',
    is_default: true,
    created_at: '2026-03-01T00:00:00Z',
    updated_at: '2026-03-01T00:00:00Z',
    ...overrides,
});

const paginatedShipments: PaginatedResponse<Shipment> = {
    success: true,
    message: 'OK',
    data: [makeShipment()],
    total: 1,
    page: 1,
    page_size: 10,
    total_pages: 1,
};

// -----------------------------------------------------------------
// Mock del repositorio
// -----------------------------------------------------------------

function createMockRepository(): IShipmentRepository {
    return {
        getShipments: vi.fn(),
        quoteShipment: vi.fn(),
        generateGuide: vi.fn(),
        trackShipment: vi.fn(),
        cancelShipment: vi.fn(),
        createShipment: vi.fn(),
        getOriginAddresses: vi.fn(),
        createOriginAddress: vi.fn(),
        updateOriginAddress: vi.fn(),
        deleteOriginAddress: vi.fn(),
    } as unknown as IShipmentRepository;
}

// -----------------------------------------------------------------
// Suite principal
// -----------------------------------------------------------------

describe('ShipmentUseCases', () => {
    let repo: ReturnType<typeof createMockRepository>;
    let useCases: ShipmentUseCases;

    beforeEach(() => {
        repo = createMockRepository();
        useCases = new ShipmentUseCases(repo as unknown as IShipmentRepository);
    });

    // ---------------------------------------------------------------
    // getShipments
    // ---------------------------------------------------------------
    describe('getShipments', () => {
        it('debería retornar la lista paginada de envíos cuando el repositorio tiene éxito', async () => {
            vi.mocked(repo.getShipments).mockResolvedValue(paginatedShipments);

            const result = await useCases.getShipments({ page: 1, page_size: 10 });

            expect(result).toEqual(paginatedShipments);
            expect(repo.getShipments).toHaveBeenCalledOnce();
            expect(repo.getShipments).toHaveBeenCalledWith({ page: 1, page_size: 10 });
        });

        it('debería llamar al repositorio sin parámetros cuando no se pasan filtros', async () => {
            vi.mocked(repo.getShipments).mockResolvedValue(paginatedShipments);

            await useCases.getShipments();

            expect(repo.getShipments).toHaveBeenCalledWith(undefined);
        });

        it('debería propagar el error cuando el repositorio falla', async () => {
            const expectedError = new Error('Fallo de base de datos');
            vi.mocked(repo.getShipments).mockRejectedValue(expectedError);

            await expect(useCases.getShipments()).rejects.toThrow('Fallo de base de datos');
        });
    });

    // ---------------------------------------------------------------
    // quoteShipment
    // ---------------------------------------------------------------
    describe('quoteShipment', () => {
        const quoteReq: EnvioClickQuoteRequest = {
            description: 'Paquete de prueba',
            contentValue: 50000,
            includeGuideCost: true,
            codPaymentMethod: 'cash',
            packages: [{ weight: 2, height: 10, width: 10, length: 10 }],
            origin: { address: 'Calle 10', daneCode: '11001' },
            destination: { address: 'Calle 20', daneCode: '05001' },
        };

        it('debería retornar cotización de envío', async () => {
            const quoteResponse: EnvioClickQuoteResponse = {
                success: true,
                message: 'OK',
                correlation_id: 'corr-123',
                data: { rates: [{ idRate: 1, idProduct: 1, product: 'Express', idCarrier: 1, carrier: 'Servientrega', flete: 15000, deliveryDays: 2, quotationType: 'standard' }] },
            };
            vi.mocked(repo.quoteShipment).mockResolvedValue(quoteResponse);

            const result = await useCases.quoteShipment(quoteReq);

            expect(result).toEqual(quoteResponse);
            expect(repo.quoteShipment).toHaveBeenCalledWith(quoteReq);
        });

        it('debería propagar el error cuando la cotización falla', async () => {
            vi.mocked(repo.quoteShipment).mockRejectedValue(new Error('Servicio de cotización no disponible'));

            await expect(useCases.quoteShipment(quoteReq)).rejects.toThrow('Servicio de cotización no disponible');
        });
    });

    // ---------------------------------------------------------------
    // generateGuide
    // ---------------------------------------------------------------
    describe('generateGuide', () => {
        const guideReq: EnvioClickQuoteRequest = {
            idRate: 1,
            carrier: 'Servientrega',
            description: 'Paquete',
            contentValue: 50000,
            includeGuideCost: true,
            codPaymentMethod: 'cash',
            packages: [{ weight: 2, height: 10, width: 10, length: 10 }],
            origin: { address: 'Calle 10', daneCode: '11001' },
            destination: { address: 'Calle 20', daneCode: '05001' },
        };

        it('debería generar una guía y retornar la respuesta', async () => {
            const generateResponse: EnvioClickGenerateResponse = {
                success: true,
                message: 'Guía generada',
                correlation_id: 'corr-456',
                shipment_id: 1,
            };
            vi.mocked(repo.generateGuide).mockResolvedValue(generateResponse);

            const result = await useCases.generateGuide(guideReq);

            expect(result).toEqual(generateResponse);
            expect(repo.generateGuide).toHaveBeenCalledWith(guideReq);
        });

        it('debería propagar el error cuando la generación de guía falla', async () => {
            vi.mocked(repo.generateGuide).mockRejectedValue(new Error('Error al generar guía'));

            await expect(useCases.generateGuide(guideReq)).rejects.toThrow('Error al generar guía');
        });
    });

    // ---------------------------------------------------------------
    // trackShipment
    // ---------------------------------------------------------------
    describe('trackShipment', () => {
        it('debería retornar el tracking de un envío', async () => {
            const trackingResponse: EnvioClickTrackingResponse = {
                success: true,
                message: 'OK',
                data: {
                    trackingNumber: 'TRACK-001',
                    carrier: 'Servientrega',
                    status: 'in_transit',
                    history: [{ date: '2026-03-01', status: 'shipped', description: 'Paquete enviado', location: 'Bogota' }],
                },
            };
            vi.mocked(repo.trackShipment).mockResolvedValue(trackingResponse);

            const result = await useCases.trackShipment('TRACK-001');

            expect(result).toEqual(trackingResponse);
            expect(repo.trackShipment).toHaveBeenCalledWith('TRACK-001');
        });

        it('debería propagar el error cuando el tracking falla', async () => {
            vi.mocked(repo.trackShipment).mockRejectedValue(new Error('Tracking no encontrado'));

            await expect(useCases.trackShipment('INVALID')).rejects.toThrow('Tracking no encontrado');
        });
    });

    // ---------------------------------------------------------------
    // cancelShipment
    // ---------------------------------------------------------------
    describe('cancelShipment', () => {
        it('debería cancelar un envío y retornar confirmación', async () => {
            const cancelResponse: EnvioClickCancelResponse = {
                success: true,
                message: 'OK',
                data: { status: 'cancelled', message: 'Envío cancelado' },
            };
            vi.mocked(repo.cancelShipment).mockResolvedValue(cancelResponse);

            const result = await useCases.cancelShipment('1');

            expect(result).toEqual(cancelResponse);
            expect(repo.cancelShipment).toHaveBeenCalledWith('1');
        });

        it('debería propagar el error cuando la cancelación falla', async () => {
            vi.mocked(repo.cancelShipment).mockRejectedValue(new Error('Envío ya entregado'));

            await expect(useCases.cancelShipment('1')).rejects.toThrow('Envío ya entregado');
        });
    });

    // ---------------------------------------------------------------
    // createShipment
    // ---------------------------------------------------------------
    describe('createShipment', () => {
        const createReq: CreateShipmentRequest = {
            order_id: 'order-1',
            client_name: 'Juan Perez',
            destination_address: 'Calle 20 #10-30',
            carrier: 'Servientrega',
            status: 'pending',
        };

        it('debería crear un envío y retornar la respuesta', async () => {
            const createResponse = { success: true, message: 'Envío creado', data: makeShipment() };
            vi.mocked(repo.createShipment).mockResolvedValue(createResponse);

            const result = await useCases.createShipment(createReq);

            expect(result).toEqual(createResponse);
            expect(repo.createShipment).toHaveBeenCalledOnce();
            expect(repo.createShipment).toHaveBeenCalledWith(createReq);
        });

        it('debería propagar el error cuando la creación falla', async () => {
            vi.mocked(repo.createShipment).mockRejectedValue(new Error('Orden no encontrada'));

            await expect(useCases.createShipment(createReq)).rejects.toThrow('Orden no encontrada');
        });
    });

    // ---------------------------------------------------------------
    // getOriginAddresses
    // ---------------------------------------------------------------
    describe('getOriginAddresses', () => {
        it('debería retornar las direcciones de origen', async () => {
            const addresses = [makeOriginAddress()];
            vi.mocked(repo.getOriginAddresses).mockResolvedValue(addresses);

            const result = await useCases.getOriginAddresses();

            expect(result).toEqual(addresses);
            expect(repo.getOriginAddresses).toHaveBeenCalledWith(undefined);
        });

        it('debería pasar businessId cuando se proporciona', async () => {
            vi.mocked(repo.getOriginAddresses).mockResolvedValue([]);

            await useCases.getOriginAddresses(5);

            expect(repo.getOriginAddresses).toHaveBeenCalledWith(5);
        });

        it('debería propagar el error cuando falla la consulta de direcciones', async () => {
            vi.mocked(repo.getOriginAddresses).mockRejectedValue(new Error('Servicio no disponible'));

            await expect(useCases.getOriginAddresses()).rejects.toThrow('Servicio no disponible');
        });
    });

    // ---------------------------------------------------------------
    // createOriginAddress
    // ---------------------------------------------------------------
    describe('createOriginAddress', () => {
        const createReq: CreateOriginAddressRequest = {
            alias: 'Nueva Bodega',
            company: 'Mi Empresa',
            first_name: 'Carlos',
            last_name: 'Lopez',
            email: 'carlos@test.com',
            phone: '3009876543',
            street: 'Carrera 5 #12-34',
            city_dane_code: '11001',
            city: 'Bogota',
            state: 'Cundinamarca',
        };

        it('debería crear una dirección de origen y retornar la respuesta', async () => {
            const newAddress = makeOriginAddress({ id: 2, alias: 'Nueva Bodega' });
            vi.mocked(repo.createOriginAddress).mockResolvedValue(newAddress);

            const result = await useCases.createOriginAddress(createReq);

            expect(result).toEqual(newAddress);
            expect(repo.createOriginAddress).toHaveBeenCalledWith(createReq, undefined);
        });

        it('debería pasar businessId cuando se proporciona', async () => {
            vi.mocked(repo.createOriginAddress).mockResolvedValue(makeOriginAddress());

            await useCases.createOriginAddress(createReq, 5);

            expect(repo.createOriginAddress).toHaveBeenCalledWith(createReq, 5);
        });

        it('debería propagar el error cuando la creación falla', async () => {
            vi.mocked(repo.createOriginAddress).mockRejectedValue(new Error('Dirección duplicada'));

            await expect(useCases.createOriginAddress(createReq)).rejects.toThrow('Dirección duplicada');
        });
    });

    // ---------------------------------------------------------------
    // updateOriginAddress
    // ---------------------------------------------------------------
    describe('updateOriginAddress', () => {
        const updateReq: UpdateOriginAddressRequest = { alias: 'Bodega Actualizada' };

        it('debería actualizar una dirección de origen y retornar la respuesta', async () => {
            const updatedAddress = makeOriginAddress({ alias: 'Bodega Actualizada' });
            vi.mocked(repo.updateOriginAddress).mockResolvedValue(updatedAddress);

            const result = await useCases.updateOriginAddress(1, updateReq);

            expect(result).toEqual(updatedAddress);
            expect(repo.updateOriginAddress).toHaveBeenCalledWith(1, updateReq, undefined);
        });

        it('debería pasar businessId cuando se proporciona', async () => {
            vi.mocked(repo.updateOriginAddress).mockResolvedValue(makeOriginAddress());

            await useCases.updateOriginAddress(1, updateReq, 5);

            expect(repo.updateOriginAddress).toHaveBeenCalledWith(1, updateReq, 5);
        });

        it('debería propagar el error cuando la actualización falla', async () => {
            vi.mocked(repo.updateOriginAddress).mockRejectedValue(new Error('Dirección no encontrada'));

            await expect(useCases.updateOriginAddress(99, updateReq)).rejects.toThrow('Dirección no encontrada');
        });
    });

    // ---------------------------------------------------------------
    // deleteOriginAddress
    // ---------------------------------------------------------------
    describe('deleteOriginAddress', () => {
        it('debería eliminar una dirección de origen y retornar confirmación', async () => {
            const deleteResponse = { message: 'Dirección eliminada' };
            vi.mocked(repo.deleteOriginAddress).mockResolvedValue(deleteResponse);

            const result = await useCases.deleteOriginAddress(1);

            expect(result).toEqual(deleteResponse);
            expect(repo.deleteOriginAddress).toHaveBeenCalledWith(1, undefined);
        });

        it('debería pasar businessId cuando se proporciona', async () => {
            vi.mocked(repo.deleteOriginAddress).mockResolvedValue({ message: 'OK' });

            await useCases.deleteOriginAddress(1, 5);

            expect(repo.deleteOriginAddress).toHaveBeenCalledWith(1, 5);
        });

        it('debería propagar la excepción cuando el repositorio lanza un error de red', async () => {
            vi.mocked(repo.deleteOriginAddress).mockRejectedValue(new Error('Network error'));

            await expect(useCases.deleteOriginAddress(1)).rejects.toThrow('Network error');
        });
    });
});
